presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java

```diff
@@ -516,5 +518,27 @@ public PinotQueryGeneratorContext visitTopN(TopNNode node, PinotQueryGeneratorCo
             checkSupported(node.getStep().equals(TopNNode.Step.SINGLE), "Can only push single logical topn in");
             return context.withTopN(getOrderingScheme(node), node.getCount()).withOutputColumns(node.getOutputVariables());
         }
+
+        @Override
+        public PinotQueryGeneratorContext visitDistinctLimit(DistinctLimitNode node, PinotQueryGeneratorContext context)
+        {
+            context = node.getSource().accept(this, context);
+            requireNonNull(context, "context is null");
+            checkSupported(!forbidBrokerQueries, "Cannot push distinctLimit in segment mode");
+            LinkedHashSet<VariableReferenceExpression> groupByColumns = new LinkedHashSet<>(node.getDistinctVariables());
+            if (!useSqlSyntax) {
+                // Handling PQL by adding hidden expression: count(*)
+                // E.g. `SELECT DISTINCT A, B FROM myTable LIMIT 10`
+                // In Pql mode, the generated pql is `SELECT count(*) FROM myTable GROUP BY A, B LIMIT 10`;
+                // In Sql mode, the generated sql is still `SELECT A, B FROM myTable GROUP BY A, B LIMIT 10`.
+                VariableReferenceExpression hidden = new VariableReferenceExpression(UUID.randomUUID().toString(), BigintType.BIGINT);
+                Map<VariableReferenceExpression, Selection> newSelections = new HashMap<>(context.getSelections());
+                newSelections.put(hidden, new Selection("count(*)", DERIVED));
+                LinkedHashSet<VariableReferenceExpression> newOutputs = new LinkedHashSet<>(groupByColumns);
+                newOutputs.add(hidden);
+                return context.withAggregation(newSelections, newOutputs, groupByColumns, 1, ImmutableSet.of(hidden)).withLimit(node.getLimit());
+            }
```

Comment:
nit: I somehow prefer having a clean else block here to denote that the SQL and PQL handling is different. I somehow missed that in my first read.

Reply:
done.